python-spec/src/somacore/spatialdata.py

```diff
@@ -0,0 +1,763 @@
+        """
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def level_count(self) -> int:
+        """The number of image levels stored in the MultiscaleImage.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def level_properties(self, level: Union[int, str]) -> "ImageProperties":
+        """The properties of an image at the specified level.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    def reference_level(self) -> Optional[int]:
+        """The index of image level that is used as a reference level.
+
+        This will return ``None`` if no current image level matches the size of the
+        reference level.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def reference_level_properties(self) -> "ImageProperties":
+        """The image properties of the reference level.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+
+class ImageProperties(Protocol):
+    """Class requirements for level properties of images.
+
+    Lifecycle: experimental
+    """
+
+    @property
+    def name(self) -> str:
+        """The key for the image.
+
+        Lifecycle: experimental
+        """
+
+    @property
+    def image_type(self) -> str:
+        """The axis order of the image data using standardized names.
+
+        A valid image type is a permuation of 'YX', 'YXC', 'YXZ', or 'YXZC'. The
+        letters have the following meanings:
+
+        * 'X' - image width
+        * 'Y' - image height
+        * 'Z' - image depth (for three dimensional images)
+        * 'C' - channels/bands
+
+        Lifecycle: experimental
+        """
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Size of each axis of the image.
+
+        Lifecycle: experimental
+        """
+
+
+@dataclass
+class SpatialRead(Generic[_ReadData]):
+    """Reader for spatial data.
+
+    Args:
+        data: The data accessor.
+        data_coordinate_space: The coordinate space the read data is defined on.
+        output_coordinate_space: The requested output coordinate space.
+        coordinate_transform: A coordinate transform from the data coordinate space to
+            the desired output coordinate space.
+
+    Lifecycle: experimental
+    """
+
+    data: _ReadData
+    data_coordinate_space: coordinates.CoordinateSpace
+    output_coordinate_space: coordinates.CoordinateSpace
+    coordinate_transform: coordinates.CoordinateTransform
+
+    def __post_init__(self):
+        if (
+            self.data_coordinate_space.axis_names
+            != self.coordinate_transform.input_axes
+        ):
+            raise ValueError(
+                "Input coordinate transform axis names do not match the data coordinate"
+                "space."
+            )
```

Comment:
Error message shows coordinatespace without space.

```suggestion
            raise ValueError(
                "Input coordinate transform axis names do not match the data coordinate "
                "space."
            )
```